File: lz77.py
def do_sequences_match(source, index1, index2, length):
    "time optimised version for comparing 2 subsequences for equality"
    terminate = index1+length

    while index1 < terminate:
        if source[index1] != source[index2]:
            return False
        index1 += 1
        index2 += 1
    return True


def find_longest_prefix_in_window(symbols, symbol_i, 
        window_size, lookahead_window_size):

    if symbol_i == 0:
        return (symbol_i, 1)

    start_of_window = max(0, symbol_i-window_size)
    end_of_lookahead = min(len(symbols), symbol_i+lookahead_window_size+1)

    longest_prefix_i, longest_prefix_len = symbol_i, 1
    window_i, cur_prefix_len = symbol_i-1, 1

    while start_of_window <= window_i:
        looking_outside_window = window_i + cur_prefix_len > symbol_i

        if looking_outside_window:
            window_i -= 1
        elif do_sequences_match(symbols, window_i, symbol_i, cur_prefix_len):
            longest_prefix_i = window_i
            longest_prefix_len = cur_prefix_len

            cur_prefix_len += 1  # going to look for a longer prefix

            looking_outside_lookahead_window = symbol_i + cur_prefix_len > end_of_lookahead
            if looking_outside_lookahead_window:
                break  # can
        else:
            window_i -= 1

    return (longest_prefix_i, longest_prefix_len)

def lz77_encode(symbols, window_size=65535, lookahead_window_size=255):
    result = []
    symbol_i = 0

    while symbol_i < len(symbols):
        (prefix_i, prefix_len) = find_longest_prefix_in_window(
            symbols, symbol_i, window_size, lookahead_window_size)

        if prefix_i == symbol_i:
            result.append((0, 0, symbols[symbol_i]))
            symbol_i += 1
        else:
            how_far_back = symbol_i - prefix_i
            next_symbol = symbols[symbol_i+prefix_len] if \
                symbol_i + prefix_len < len(symbols) else None

            result.append((how_far_back, prefix_len, next_symbol))
            symbol_i += prefix_len+1

    return result


def lz77_decode(symbols):
    "Decodes a LZ77 encoded sequence of symbols"
    result, ptr = [], 0
    for (far_back, length, symbol) in symbols:
        if length > 0:
            result.extend(result[ptr-far_back:ptr-far_back+length])
        if symbol != None:
            result.append(symbol)
        ptr += length+1

    return result


def string_to_bytes(str):
    "Converts a string into an array of bytes."
    b = bytearray()
    b.extend(map(ord, str))
    return b


def lz77_encode_string(string, window_size=65535, lookahead_window_size=255):
    "Encodes string using LZ77 coding."
    return lz77_encode(
        string_to_bytes(string), window_size, lookahead_window_size)


def lz77_decode_string(source):
    "Decodes LZ77 encoded string."
    return ''.join(chr(x) for x in lz77_decode(source))


def check_string(string):
    "asserts that string can be encoded and decoded using LZ77 coding"
    matches = string == lz77_decode_string(lz77_encode_string(string))
    if not matches:
        print("failed on string:", string)
        assert False

File: LZ77.java
package coders;

import java.util.List;
import java.util.ArrayList;

import coders.LZ77Triple;
import coders.EncodedMessage;
import coders.LZ77CodedMessage;

class FatIndex {
    public final int index;
    public final int len;

    public FatIndex(int index, int len) {
        this.index = index;
        this.len = len;
    }
}

public class LZ77 {
    private final int window, lookahead;

    private static final byte NO_SYMBOL = -1;

    /**
     * Decodes an LZ77 encoded message.
     * 
     * @param triples List to decode
     * @return Decoded message
     */
    public static byte[] decode(LZ77CodedMessage encodedMessage) {
        var bufferSize = encodedMessage.getDecodedSize();
        var buffer = new byte[bufferSize];
        var ptr = 0;

        for (int n = 0; n < encodedMessage.getNumberOfTriples(); ++n) {
            var triple = encodedMessage.getTriple(n);

            if (triple.l > 0) {
                var end = ptr - triple.d + triple.l;
                for (var i = ptr - triple.d; i < end; ++i, ++ptr) {
                    buffer[ptr] = buffer[i];
                }
            }

            if (triple.next != NO_SYMBOL) {
                buffer[ptr] = triple.next;
                ptr += 1;
            }
        }

        return buffer;
    }

    private boolean doSequencesMatch(byte[] symbols, int index1, int index2, int length) {
        var termianteIndex = index1 + length;

        for (; index1 < termianteIndex; ++index1, ++index2) {
            if (symbols[index1] != symbols[index2]) {
                return false;
            }
        }

        return true;
    }

    private FatIndex findLongestPrefixInWindow(byte[] symbols, int symbolI) {
        if (symbolI == 0) {
            return new FatIndex(symbolI, 0);
        }

        final int startOfWindow = Math.max(0, symbolI - window);
        final int endOfLookahead = Math.min(symbols.length, symbolI + lookahead + 1);

        int longestPrefixI = symbolI, longestPrefixLen = 1;
        int windowI = symbolI - 1, curPrefixLen = 1;

        while (startOfWindow <= windowI) {
            var lookingOutsideWindow = windowI + curPrefixLen > symbolI;
            if (lookingOutsideWindow) {
                --windowI;
            } else if (doSequencesMatch(symbols, windowI, symbolI, curPrefixLen)) {
                longestPrefixI = windowI;
                longestPrefixLen = curPrefixLen;

                ++curPrefixLen;

                var lookingOutsideLookahead = symbolI + curPrefixLen > endOfLookahead;
                if (lookingOutsideLookahead) {
                    break;
                }
            } else {
                --windowI;
            }
        }

        return new FatIndex(longestPrefixI, longestPrefixLen);
    }

    /**
     * Encodes an array of symbols
     * 
     * @param symbols symbols to encode
     * @return encoded symbols
     */
    public LZ77CodedMessage encode(byte[] symbols) {
        var triples = new ArrayList<LZ77Triple>();

        for (int symbolI = 0; symbolI < symbols.length; ++symbolI) {
            var prefix = findLongestPrefixInWindow(symbols, symbolI);

            if (prefix.index == symbolI) {
                triples.add(new LZ77Triple(0, (char) 0, symbols[symbolI]));
            } else {
                var howFarBack = symbolI - prefix.index;
                var nextSymbolI = symbolI + prefix.len;
                var nextSymbol = (nextSymbolI < symbols.length) ? symbols[nextSymbolI] : -1;

                triples.add(new LZ77Triple(howFarBack, (char) prefix.len, nextSymbol));
                symbolI += prefix.len;
            }
        }

        return new LZ77CodedMessage(triples);
    }

    /**
     * Encodes a given string
     * 
     * @param s string to encode
     * @return encoded string
     */
    public LZ77CodedMessage encode(String s) {
        return encode(s.getBytes());
    }

    /**
     * Create LZ77 with window size 65535 and lookahead size 255
     */
    public LZ77() {
        this(65535, 255);
    }

    /**
     * Create new LZ77 coder with specific window and lookahead sizes.
     */
    public LZ77(int window, int lookahead) {
        this.window = window;
        this.lookahead = lookahead;
    }
}
